crates/red_knot/src/symbols.rs

```diff
@@ -545,28 +546,32 @@ where
 #[derive(Debug)]
 pub(crate) struct ReachableDefinitionsIterator<'a> {
     table: &'a SymbolTable,
-    flow_node_id: FlowNodeId,
     symbol_id: SymbolId,
+    pending: Vec<FlowNodeId>,
 }
 
 impl<'a> Iterator for ReachableDefinitionsIterator<'a> {
     type Item = Definition;
 
     fn next(&mut self) -> Option<Self::Item> {
-        loop {
-            match &self.table.flow_graph.flow_nodes_by_id[self.flow_node_id] {
-                FlowNode::Start => return None,
+        while let Some(flow_node_id) = self.pending.pop() {
+            match &self.table.flow_graph.flow_nodes_by_id[flow_node_id] {
+                FlowNode::Start => return Some(Definition::None),
                 FlowNode::Definition(def_node) => {
                     if def_node.symbol_id == self.symbol_id {
-                        // we found a definition; previous definitions along this path are not
-                        // reachable
-                        self.flow_node_id = FlowGraph::start();
                         return Some(def_node.definition.clone());
                     }
-                    self.flow_node_id = def_node.predecessor;
+                    self.pending.push(def_node.predecessor);
+                }
+                FlowNode::Branch(branch_node) => {
+                    self.pending.push(branch_node.predecessor);
+                }
+                FlowNode::Phi(phi_node) => {
+                    self.pending.extend_from_slice(&phi_node.predecessors);
```

Comment:
Does the order that we explore the basic blocks prior to a phi node matter? I guess it's going to just be whatever order they are within `.predecessors`

Reply:
In principle it shouldn't matter; in practice it may be more intuitive if, when we synthesize a union type from multiple reachable definitions, we try to display that union with elements in source order. I can look at this.

Reply:
I wonder if the text-locations on the AST nodes can be included in these predecessors, and then instead of a vec we could store them as a heap. That would allow you to add them in any order but obtain source order.

Reply:
I think we could do that, but we can also achieve the same result just by being careful about the order we list predecessors in a phi node and push them to the pending list in the flow graph.

Reply:
Got this working and pushed.

Reply:
I was wrong here (inadequate test cases.) It's not generally possible to do this the way I was trying to do it (statically in the structure of the flow graph), since some names may be defined in branches where others aren't. If we really want to guarantee source order of inferred unions, we would have to explicitly include source-order information in definitions. Not going to prioritize that right now, though.